readthedocs/core/resolver.py

```diff
@@ -214,6 +214,58 @@ def resolve(
         )
         return urlunparse((protocol, domain, path, '', query_params, ''))
 
+    def _get_root_path(self, project):
+        """
+        Returns the root path for a project.
+
+        If the project is a subproject, it will return ``/projects/<project-slug>/``.
+        If the project is a main project, it will return ``/``.
+        This will respect the custom urlconf of the project if it's defined.
+        """
+        custom_prefix = None
+        if project.urlconf:
+            custom_prefix = project.urlconf.split("$", 1)[0]
+
+        if project.is_subproject:
+            prefix = custom_prefix or "projects"
+            return join_url_path(prefix, project.slug, "/")
+
+        prefix = custom_prefix or "/"
+        return join_url_path(prefix, "/")
+
+    def resolve_root(self, project, external_version_slug=None):
+        """
+        Get the root URL from where the documentation of ``project`` is served from.
+
+        This doesn't include the version or language. For example:
+
+        - https://docs.example.com/projects/<project-slug>/
+        - https://docs.readthedocs.io/
+
+        This will respect the custom urlconf of the project if it's defined.
+
+        :param project: Project object to get the root URL from
+        :param external_version_slug: If given, resolve using the external version domain.
+        """
+        canonical_project = self._get_canonical_project(project)
+        use_custom_domain = self._use_cname(canonical_project)
+        custom_domain = canonical_project.get_canonical_custom_domain()
+        if external_version_slug:
+            domain = self._get_external_subdomain(
+                canonical_project, external_version_slug
+            )
+            use_https = settings.PUBLIC_DOMAIN_USES_HTTPS
```

Comment:
I wonder if we want to get rid of this setting.. might make things cleaner, though I guess would work less well in dev, etc.

Reply:
yeah, this is useful for dev